Optimized sql statement

steffzt opened the pull request:
in self.add_hierarchy_path_sql

as we have quite some entries in the relation table, the old
statement with the subselect took about 2 minutes to complete,
the optimized statement goes a bit faster, because the database
can use indexes reasonable again-. (0.03 ms or so)

Backlog is usable again.

geändert:       app/models/relation/hierarchy_paths.rb

steffzt committed:
Optimized sql statement

in self.add_hierarchy_path_sql

as we have quite some entries in the relation table, the old
statement with the subselect took about 2 minutes to complete,
the optimized statement goes a bit faster, because the database
can use indexes reasonable again-. (0.03 ms or so)

Backlog is usable again.

geändert:       app/models/relation/hierarchy_paths.rb

oliverguenther commented:
Hi @steffzt , thanks for your efforts. This is very appreciated. However, the statement as it is is breaking tests on MySQL:

```

Mysql2::Error: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'WHERE a.to_id=
              AND a.hierarchy >  0 AND a.relates = 0 AND a.blocks' at line 8:         INSERT INTO hierarchy_paths (work_package_id, path) 
        SELECT                              to_id, GROUP_CONCAT(from_id ORDER BY hierarchy DESC SEPARATOR ',') AS path
        FROM (
          SELECT  a.to_id, a.from_id, a.hierarchy
            FROM relations AS a
            JOIN relations AS b
            ON a.to_id = b.from_id or b.to_id = 
            WHERE a.to_id=
              AND a.hierarchy >  0 AND a.relates = 0 AND a.blocks = 0 AND a.duplicates = 0 AND a.includes = 0 AND a.requires = 0 AND a.follows = 0
              AND b.hierarchy != 0 AND b.relates = 0 AND b.duplicates = 0 AND b.follows = 0 AND b.blocks = 0 AND b.includes = 0 AND b.requires = 0
          ) tmp
           GROUP BY  to_id
        ON DUPLICATE KEY
         UPDATE hierarchy_paths.path = VALUES(path)
``` 

You can see the entire run output at https://travis-ci.org/opf/openproject/builds/454978998?utm_source=github_status&utm_medium=notification

Best,
Oliver

steffzt committed:
Handle id.nil?

geändert:       app/models/relation/hierarchy_paths.rb

steffzt committed:
Removed trailing whitespaces...

geändert:       app/models/relation/hierarchy_paths.rb

steffzt committed:
Avoid duplicate entries in hierarchy_paths.path

geändert:       app/models/relation/hierarchy_paths.rb

steffzt committed:
Updated add_hierarchy_path_sql

Request now a stach of union selects; much faster now, and rspec
test pass.

app/models/relation/hierarchy_paths.rb

steffzt committed:
Added branch: 'v1.0.3' to rspec-activemodel-mocks

The current master on gitub is 1.1.0 ans 1.0.3 is
not available there

geändert:       Gemfile
geändert:       Gemfile.lock

steffzt committed:
AWS plugins stopped workin

-gem 'aws-sdk-core', '~> 3.20.2'
+gem 'aws-sdk-core', '~> 3.38.0' #'~> 3.20.2'
 # File upload via fog + screenshots on travis
-gem 'aws-sdk-s3', '~> 1.9.1'
+gem 'aws-sdk-s3', '~> 1.24.0' #, '~> 1.9.1'

and try to make bundle run again, wich did.

geändert:       Gemfile
geändert:       Gemfile.lock

steffzt committed:
Fixed code climate complaints

geändert:       Gemfile
geändert:       app/models/relation/hierarchy_paths.rb

steffzt committed:
Merge branch 'dev' of https://github.com/opf/openproject into dev

steffzt committed:
Merge branch 'dev' into feature/optimize-hierarchy-sql

steffzt committed:
Merge branch 'dev' of https://github.com/opf/openproject into dev

steffzt committed:
Merge branch 'dev' into feature/optimize-hierarchy-sql

steffzt committed:
Merge branch 'dev' into feature/optimize-hierarchy-sql

steffzt committed:
Merge branch 'dev' into feature/optimize-hierarchy-sql

steffzt committed:
Unfreeze WHITELIST fro modification

and freeze again after done.
This ignores the warning about assigning to an already assigned constant.

geändert:       lib/open_project/text_formatting/filters/sanitization_filter.rb

steffzt committed:
Follow code climete recommendations

geändert:       Gemfile
geändert:       lib/open_project/text_formatting/filters/sanitization_filter.rb

steffzt committed:
again

steffzt committed:
removed typo in Gemfile

steffzt committed:
FactoryBot/AttributeDefinedStatically

rubocop pt all into blocks

	geändert:       modules/avatars/spec/factories/avatar_attachment_factory.rb
	geändert:       modules/backlogs/spec/factories/impediment_factory.rb
	geändert:       modules/backlogs/spec/factories/sprint_factory.rb
	geändert:       modules/backlogs/spec/factories/story_factory.rb
	geändert:       modules/backlogs/spec/factories/task_factory.rb
	geändert:       modules/costs/spec/factories/cost_entry_factory.rb
	geändert:       modules/costs/spec/factories/cost_object_factory.rb
	geändert:       modules/costs/spec/factories/cost_rate_factory.rb
	geändert:       modules/costs/spec/factories/cost_type_factory.rb
	geändert:       modules/costs/spec/factories/default_hourly_rate_factory.rb
	geändert:       modules/costs/spec/factories/hourly_rate_factory.rb
	geändert:       modules/costs/spec/factories/labor_budget_item_factory.rb
	geändert:       modules/costs/spec/factories/material_budget_item_factory.rb
	geändert:       modules/costs/spec/factories/rate_factory.rb
	geändert:       modules/costs/spec/factories/variable_cost_object_factory.rb
	geändert:       modules/ldap_groups/spec/factories/synchronized_group_factory.rb
	geändert:       modules/meeting/spec/factories/meeting_journal_factory.rb
	geändert:       modules/pdf_export/spec/factories/export_card_configuration_factory.rb
	geändert:       modules/reporting/spec/factories/cost_query_factory.rb
	geändert:       modules/two_factor_authentication/spec/factories/two_factor_authentication_device_factory.rb
	geändert:       modules/webhooks/spec/factories/webhook_factory.rb
	geändert:       modules/webhooks/spec/factories/webhook_log_factory.rb
	geändert:       spec/factories/announcement_factory.rb
	geändert:       spec/factories/attachment_factory.rb
	geändert:       spec/factories/attribute_help_text_factory.rb
	geändert:       spec/factories/auth_source_factory.rb
	geändert:       spec/factories/changeset_factory.rb
	geändert:       spec/factories/color_factory.rb
	geändert:       spec/factories/custom_field_factory.rb
	geändert:       spec/factories/custom_value_factory.rb
	geändert:       spec/factories/design_color_factory.rb
	geändert:       spec/factories/enumerations_factory.rb
	geändert:       spec/factories/file_factory.rb
	geändert:       spec/factories/issue_priority_factory.rb
	geändert:       spec/factories/journal/wiki_content_journal_factory.rb
	geändert:       spec/factories/journal_factory.rb
	geändert:       spec/factories/menu_item_factory.rb
	geändert:       spec/factories/principal_factory.rb
	geändert:       spec/factories/project_factory.rb
	geändert:       spec/factories/query_factory.rb
	geändert:       spec/factories/relation_factory.rb
	geändert:       spec/factories/repository_factory.rb
	geändert:       spec/factories/role_factory.rb
	geändert:       spec/factories/status_factory.rb
	geändert:       spec/factories/time_entry_factory.rb
	geändert:       spec/factories/type_factory.rb
	geändert:       spec/factories/user_factory.rb
	geändert:       spec/factories/user_password_factory.rb
	geändert:       spec/factories/user_preference_factory.rb
	geändert:       spec/factories/version_factory.rb
	geändert:       spec/factories/wiki_factory.rb
	geändert:       spec/factories/wiki_redirect_factory.rb
	geändert:       spec/factories/work_package_custom_field_factory.rb
	geändert:       spec/factories/work_package_factory.rb

steffzt committed:
Merge branch 'dev' into feature/optimize-hierarchy-sql

steffzt committed:
Merge branch 'dev' into feature/optimize-hierarchy-sql

ulferts commented:
@steffzt please see #6847 for a review. If you like to have this PR merged, and by that your great work attributed to you, please address the points mentioned there and then squash the commits.
